File: examples/lennardjones/ljcalculator.py
#!/usr/bin/env python

"""Demonstration of using PairQuantity class for calculation of Lennard Jones
potential.

Vij = 4 * ( rij ** -12  -  rij ** -6 )
"""

import sys

from diffpy.srreal.pairquantity import PairQuantity
from diffpy.structure import Structure


class LennardJonesCalculator(PairQuantity):

    # Initialization.  The size of the result array is always 1.
    def __init__(self):
        PairQuantity.__init__(self)
        self._resizeValue(1)
        return

    def __call__(self, structure):
        """Return LJ potential for a specified structure."""
        values = self.eval(structure)
        return values[0]

    def _addPairContribution(self, bnds, sumscale):
        """Add Lennard-Jones contribution from a single pair of atoms.

        bnds     -- a BaseBondGenerator instance that contains information
                    about the current pair of atom sites.
        sumscale -- integer multiplicity of the current pair.  It may
                    be negative in case of fast value updates.

        No return value.  Updates _value, the internal results array.
        """
        rij = bnds.distance()
        ljij = 4 * (rij**-12 - rij**-6)
        self._value[0] += sumscale * ljij / 2.0
        return


# class LennardJonesCalculator


def main():
    # load structure from a specified file, by default "lj50.xyz"
    filename = len(sys.argv) > 1 and sys.argv[1] or "lj50.xyz"
    stru = Structure()
    stru.read(filename)
    # create an instance of LennardJonesCalculator
    ljcalc = LennardJonesCalculator()
    # calculate and print the LJ potential.
    print("LJ potential of %s is %g" % (filename, ljcalc(stru)))


if __name__ == "__main__":
    main()

File: docs/examples/lennardjones/ljcalculator.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <diffpy/srreal/PairQuantity.hpp>
#include <diffpy/srreal/VR3Structure.hpp>

using namespace diffpy::srreal;
using namespace std;

class LennardJonesCalculator : public PairQuantity
{
    public:
        // Constructor - the size of the results array is always one,
        // because LJ potential is a scalar
        LennardJonesCalculator()
        {
            this->resizeValue(1);
        }

        // convenience template method.  This allows using an instantiated
        // LennardJonesCalculator as a function.  T must be a type for which
        // a createStructureAdapter(const T&) function exists.
        template <class T>
        double operator()(const T& stru)
        {
            this->eval(stru);
            return this->value()[0];
        }

    protected:

        // define contribution from a pair of atoms.  mvalue is a reference
        // to the internal array of result values.
        void addPairContribution(const BaseBondGenerator& bnds, int sumscale)
        {
            double rij = bnds.distance();
            double ljij = 4 * (pow(rij, -12) - pow(rij, -6));
            mvalue[0] += sumscale * ljij / 2.0;
        }

};


// helper function that loads a simple list of xyz coordinates
VR3Structure loadVR3Structure(string filename)
{
    VR3Structure rv;
    ifstream fp(filename.c_str());
    double x, y, z;
    while (fp >> x >> y >> z)
    {
        rv.push_back(R3::Vector(x, y, z));
    }
    return rv;
}


int main(int argc, char* argv[])
{
    using namespace std;
    string filename;
    filename = (argc > 1) ? argv[1] : "lj50.xyz";
    LennardJonesCalculator ljcalc;
    VR3Structure stru = loadVR3Structure(filename);
    cout << "LJ potential of " << filename << " is " << ljcalc(stru) << '\n';
    return 0;
}
